File: ctci/python/ch1-arrays-and-strings/q2-check-permutation/QuestionA.py
from collections import defaultdict


# assume ASCII
def check_permutation(s1, s2):
    if len(s1) != len(s2):
        return False

    counts = [0 for _ in xrange(128)]
    for i in xrange(len(s1)):
        counts[ord(s1[i])] += 1
        counts[ord(s2[i])] -= 1
    for count in counts:
        if count != 0:
            return False
    return True


# no assumption
def check_permutation_no_assumption(s1, s2):
    if len(s1) != len(s2):
        return False

    counts = defaultdict(int)
    for i in xrange(len(s1)):
        counts[s1[i]] += 1
        counts[s2[i]] -= 1
    for count in counts.values():
        if count != 0:
            return False
    return True

File: ctci/java/ch1-arrays-and-strings/q2_check_permutation/QuestionA.java
package q2_check_permutation;

import java.util.HashMap;

/**
 * Created by jjlee on 7/10/16.
 */
public class QuestionA {
    /* assume ASCII */
    public static boolean checkPermutation(String s1, String s2) {
        if (s1.length() != s2.length())
            return false;

        int[] counts = new int[128];
        for (int i = 0; i < s1.length(); i++) {
            char c1 = s1.charAt(i);
            char c2 = s2.charAt(i);
            counts[c1]++;
            counts[c2]--;
        }
        for (int count : counts) {
            if (count != 0)
                return false;
        }
        return true;
    }

    /* no assumption */
    public static boolean checkPermutationNoAssumption(String s1, String s2) {
        if (s1.length() != s2.length())
            return false;

        HashMap<Character, Integer> counts = new HashMap<>();
        for (int i = 0; i < s1.length(); i++) {
            char c1 = s1.charAt(i);
            char c2 = s2.charAt(i);
            counts.put(c1, (counts.containsKey(c1) ? counts.get(c1) : 0) + 1);
            counts.put(c2, (counts.containsKey(c2) ? counts.get(c2) : 0) - 1);
        }
        for (char c : counts.keySet()) {
            if (counts.get(c) != 0)
                return false;
        }
        return true;
    }
}
